qiskit_experiments/calibration/experiments/rabi.py

```diff
@@ -0,0 +1,337 @@
+            freq_bounds=(0, np.inf),
+            offset_bounds=(0, 1),
+            phase_bounds=(-np.pi, np.pi),
+        )
+
+    def _run_analysis(
+        self,
+        experiment_data: ExperimentData,
+        data_processor: Optional[Callable] = None,
+        amp_guess: float = 1.0,
+        freq_guesses: List[float] = np.linspace(0, 5 * np.pi, 10),
+        offset_guess: float = 0.0,
+        phase_guess: Optional[float] = None,
+        amp_bounds: Tuple[float, float] = (-1, 1),
+        freq_bounds: Tuple[float, float] = (0, np.inf),
+        offset_bounds: Tuple[float, float] = (-1, 1),
+        phase_bounds: Tuple[float, float] = (-np.pi, np.pi),
+        plot: bool = True,
+        ax: Optional["AxesSubplot"] = None,
+    ) -> Tuple[AnalysisResult, List["plotting.pyplot.Figure"]]:
+        """Fit the data to an oscillating function.
+
+        Args:
+            experiment_data: The experiment data to fit.
+            data_processor: A data processor with which to analyse the data. If None is given
+                a SVD-based data processor will be used for kerneled data while a conversion
+                from counts to probabilities will be used for discriminated data.
+            amp_guess: The amplitude guess for the fit which will default to 0.5.
+            freq_guesses: The frequency guesses for the fit which defaults to a list of 20 points
+                linearly spaced between 0 and 2pi.
+            offset_guess: The y-axis offset which defaults to 0.5.
+            phase_guess: Phase of the oscillation which defaults to both 0 and pi.
+            amp_bounds: Bounds on the amplitude which default to (-1, 1).
+            freq_bounds: Bounds on the frequency which default to (0, inf).
+            offset_bounds: Bounds on the offset which default to (0,1).
+            phase_bounds: Bounds on the phase of the cosine which default to (-pi, pi).
+            plot: If True generate a plot of fitted data.
+            ax: Optional, matplotlib axis to add plot to.
+
+        Returns:
+            The analysis result with the fit and optional plots.
+
+        Raises:
+            QiskitError: If the fit fails.
+        """
+
+        meas_level = experiment_data.data(0)["metadata"]["meas_level"]
+        meas_return = experiment_data.data(0)["metadata"]["meas_return"]
+
+        # Pick a data processor.
+        if data_processor is None:
+            data_processor = get_to_signal_processor(meas_level=meas_level, meas_return=meas_return)
+            data_processor.train(experiment_data.data())
+
+        y_sigmas = np.array([data_processor(datum) for datum in experiment_data.data()])
+        min_y, max_y = min(y_sigmas[:, 0]), max(y_sigmas[:, 0])
+        ydata = (y_sigmas[:, 0] - min_y) / (max_y - min_y)
+
+        # Sigmas may be None and fitting will not work if any sigmas are exactly 0.
+        try:
+            sigmas = y_sigmas[:, 1] / (max_y - min_y)
+            if any(sigmas == 0.0):
```

Comment:
Could we move this logic to a core location, maybe in data processing, in a way that it will automatically apply to all the experiments?

Reply:
See PR #46